src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,601 @@
+            if (stack.Count > 0 && stack[stack.Count - 1].Multiple)
+            {
+                return;
+            }
+
+            stack.Add(new PendingOperation() { CapturedContext = knownContext, Multiple = true });
+        }
+
+        public static JSProxyContext PushOperationWithCurrentThreadContext()
+        {
+            var stack = OperationStack;
+            // for SchedulePopOperation()
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+
+            var current = AssertCurrentContext();
+            stack.Add(new PendingOperation { CapturedContext = current });
+            return current;
+        }
+
+        public static void PopOperation()
+        {
+            var stack = OperationStack;
+            stack.RemoveAt(stack.Count - 1);
+        }
+
+        // this is here until we change the code generator and the API to Push/Pop the context
+        // we have no way how to Pop from the after marshaling the result
+        public static void SchedulePopOperation()
+        {
+            var stack = OperationStack;
+            stack[stack.Count - 1].Called = true;
+        }
+
+        public static void AssertOperationStack(int expected)
+        {
+            var stack = OperationStack;
+            if (stack.Count > 0 && stack[stack.Count - 1].Called)
+            {
+                PopOperation();
+            }
+            var actual = stack.Count;
+            var multiple = stack.Count > 0 ? stack[stack.Count - 1].Multiple : false;
+            var called = stack.Count > 0 ? stack[stack.Count - 1].Called : false;
+            if (expected == 0)
+            {
+                stack.Clear();
+            }
+            // TODO Environment.FailFast
+            if (actual != expected) throw new InvalidOperationException($"Unexpected OperationStack size expected: {expected} actual: {actual} called:{called} multiple:{multiple}");
+        }
+
+        // TODO: sort generated ToJS() calls to make the capture context before we need to use it
+        public static void CaptureContextFromParameter(JSProxyContext parameterContext)
+        {
+            var stack = OperationStack;
+            var pendingOperation = stack[stack.Count - 1];
+            var capturedContext = pendingOperation.CapturedContext;
+            if (capturedContext != null && parameterContext != capturedContext)
+            {
+                throw new InvalidOperationException("All JSObject proxies need to have same thread affinity");
+            }
+            pendingOperation.CapturedContext = capturedContext;
+        }
+
+        // Context of the current thread or async task. Could be null on threads which don't have JS interop, like managed thread pool threads.
+        // TODO flow it also with ExecutionContext to child threads ?
+        private static readonly AsyncLocal<JSProxyContext?> _currentThreadContext = new AsyncLocal<JSProxyContext?>();
```

Comment:
we should find an expert to tell us whether to use asynclocal or threadlocal here. it's not clear to me which is correct

Reply:
I separated `ExecutionContext` and `CurrentThreadContext`
- `CurrentThreadContext` is good to know that this is really the thread that has JS state and interop
- `ExecutionContext` is good to provide better default target in the fallback hierarchy, when we don't have any JSObject parameter context capture.